### ISSUE TITLE
rc_buffer lint is dangerous as it changes runtime behaviour

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#6170]
clippy wants if one tries to use a type like `Rc<Vec<T>>`, .e.g

```
Arc<Vec<u8>>
^^^^^^^^^^^^ help: try: `Arc<[u8]>`
```

This is dangerous as creating a `Rc<[T]>` causes a `memcpy` of the whole memory area with all the elements, while creating an `Rc<Vec<T>>` simply copies over the 3 pointers inside the `Vec` without copying all the elements.

A better suggestion would probably be `Rc<Box<[T]>>` as that has the same runtime behaviour, but even independent of that it can be useful to have a `Vec` inside an `Rc` / `Arc` in combination with the `make_mut()` / `get_mut()` API.

The lint should probably be at least disabled by default.

----

The same also applies to `Rc<String>` vs `Rc<str>`.

**Comment by Kixunil** [comment]
In the specific case of `[u8]`, one might want to use `bytes::BytesMut` instead of `Vec<u8>`.

**Comment by cmyr** [comment]
Will second this, I think there are very reasonable cases to use `Rc<Vec<T>>` along with `Rc::make_mut`.

**Comment by GuillaumeGomez** [comment]
Just tried:

```rust
use std::sync::Arc;

struct Bar {
    inner: Arc<Vec<u8>>
}

fn foo(x: Arc<Vec<u8>>) {
    let x = Bar { inner: x };
}

fn main() {
    let x: Arc<Vec<u8>> = Arc::new(Vec::new());
    foo(x);
}
```

The lint was not emitted so I suppose we can close this issue?

**Comment by sdroege** [comment]
Would be good to understand which change caused this, but yes looks like this can be closed then :)

**Comment by y21** [comment]
It doesn't lint on the code anymore because it was moved to restriction in #6128, so it's opt in now. With `-Wclippy::rc_buffer` it still emits a warning. Which I guess is enough to close this? It's not really a false positive that can be fixed in the lint I don't think, so restriction seems like a reasonable category.

> A better suggestion would probably be `Rc<Box<[T]>>`

IMO this is worse than `Rc<[T]>` as a general suggestion. Converting a `Vec<T>` to `Box<[T]>` is also not free and involves memcpy-ing the data into a new buffer when `capacity != len`. When this happens, you have double indirection on every element access and you still need to copy the data

**Comment by samueltardieu** [comment]
I can still reproduce the issue.

@rustbot label -S-fixed

**Comment by SzilvasiPeter** [comment]
> I can still reproduce the issue.
> 
> [@rustbot](https://github.com/rustbot) label -S-fixed

@samueltardieu Could you share the code that reproduces this issue?

**Comment by samueltardieu** [comment]
The code is here already: https://github.com/rust-lang/rust-clippy/issues/6170#issuecomment-1900179813

You just have to add `#![warn(clippy::rc_buffer)]` since the lint is now allow-by-default.

**Comment by SzilvasiPeter** [comment]
> The code is here already: [#6170 (comment)](https://github.com/rust-lang/rust-clippy/issues/6170#issuecomment-1900179813)
> 
> You just have to add `#![warn(clippy::rc_buffer)]` since the lint is now allow-by-default.

I added the warn attribute, but I still don't get any warning. What am I doing wrong in [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7903b7fa1ae05f9c9448876d28d9be71)?

**Comment by samueltardieu** [comment]
> > The code is here already: [#6170 (comment)](https://github.com/rust-lang/rust-clippy/issues/6170#issuecomment-1900179813)
> > You just have to add `#![warn(clippy::rc_buffer)]` since the lint is now allow-by-default.
> 
> I added the warn attribute, but I still don't get any warning. What am I doing wrong in [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7903b7fa1ae05f9c9448876d28d9be71)?

Why don't you try with the code I pointed to?
